### PR TITLE
Extract safekeeper per timeline state from safekeeper.rs

### PR DESCRIPTION
**Pull request by arssher**

### CODE REVIEW

File: safekeeper/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+use crate::{
+    control_file,
+    safekeeper::{AcceptorState, PersistedPeerInfo, PgUuid, ServerInfo, TermHistory},
+};
+
+/// Persistent information stored on safekeeper node
+/// On disk data is prefixed by magic and format version and followed by checksum.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SafeKeeperPersistentState {
```
Comment (petuhovskiy) [inline safekeeper/src/state.rs:21]: Let's rename it to just `PersistentState`

File: safekeeper/src/state.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            Lsn::INVALID,
+            Lsn::INVALID,
+        )
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+// In memory safekeeper state. Fields mirror ones in `SafeKeeperPersistentState`; values
+// are not flushed yet.
+pub struct SafekeeperMemState {
```
Comment (petuhovskiy) [inline safekeeper/src/state.rs:120]: Let's rename it to just `MemState`

File: safekeeper/src/safekeeper.rs [hunk 3]
```diff
@@ -863,13 +753,13 @@ where
             // NB: on new clusters, this happens at the same time as
             // timeline_start_lsn initialization, it is taken outside to provide
             // upgrade.
-            self.inmem.commit_lsn = max(self.inmem.commit_lsn, state.timeline_start_lsn);
+            state.commit_lsn = max(self.state.inmem.commit_lsn, state.timeline_start_lsn);
```
Comment (petuhovskiy) [inline safekeeper/src/safekeeper.rs:756]:
```suggestion
            state.commit_lsn = max(state.commit_lsn, state.timeline_start_lsn);
```

File: safekeeper/src/safekeeper.rs [hunk 4]
```diff
@@ -863,13 +753,13 @@ where
...
 
             // Initializing backup_lsn is useful to avoid making backup think it should upload 0 segment.
-            self.inmem.backup_lsn = max(self.inmem.backup_lsn, state.timeline_start_lsn);
+            state.backup_lsn = max(self.state.inmem.backup_lsn, state.timeline_start_lsn);
```
Comment (petuhovskiy) [inline safekeeper/src/safekeeper.rs:759]:
```suggestion
            state.backup_lsn = max(state.backup_lsn, state.timeline_start_lsn);
```

**Review by petuhovskiy — APPROVED** [review]

File: safekeeper/src/state.rs (hunk 1 above)
Comment (arssher) [reply safekeeper/src/state.rs:21]: Maybe `TimelinePersistentState`? A bit too generic otherwise.

Comment (petuhovskiy) [reply safekeeper/src/state.rs:21]: Works for me, but I'm more a fan of a short names, like `state::Persistent` :)

Comment (arssher) [reply safekeeper/src/state.rs:21]: There is also SafekeeperState, and making it just State is too much, so renamed all to Timeline*

File: safekeeper/src/safekeeper.rs (hunk 3 above)
Comment (arssher) [reply safekeeper/src/safekeeper.rs:756]: done

File: safekeeper/src/safekeeper.rs (hunk 4 above)
Comment (arssher) [reply safekeeper/src/safekeeper.rs:759]: done

File: safekeeper/src/state.rs (hunk 2 above)
Comment (arssher) [reply safekeeper/src/state.rs]: done

**Pull request merged**